src/utils.jl

```diff
@@ -334,31 +336,48 @@ Base.IndexStyle(::Type{ReversedBuf}) = Base.IndexLinear()
 Base.getindex(a::ReversedBuf, i::Int) = a.buf[end + 1 - i]
 
 # multithreading structures
-const AtomicTypes = Union{
-    Bool, Int8, Int16, Int32, Int64, Int128,
-    UInt8, UInt16, UInt32, UInt64, UInt128,
-    Float16, Float32, Float64
-}
+const AtomicTypes = Union{Base.Threads.atomictypes...}
 
 struct AtomicVector{T} <: AbstractArray{T, 1}
-    A::Vector{Threads.Atomic{T}}
+    A::Vector{T}
 
     function AtomicVector{T}(undef, n::Int) where {T}
         T <: AtomicTypes || throw(ArgumentError("invalid type for AtomicVector: $T"))
-        return new{T}(map(x->Threads.Atomic{T}(), 1:n))
+        return new{T}(zeros(T, n))
     end
 
     function AtomicVector(A::Vector{T}) where {T}
         T <: AtomicTypes || throw(ArgumentError("invalid type for AtomicVector: $T"))
-        return new{T}(map(x->Threads.Atomic{T}(A[x]), 1:length(A)))
+        return new{T}(A)
     end
 end
 
 Base.size(a::AtomicVector) = size(a.A)
 Base.IndexStyle(::Type{A}) where {A <: AtomicVector} = Base.IndexLinear()
 
-Base.getindex(a::AtomicVector, i::Int) = a.A[i][]
-Base.setindex!(a::AtomicVector, x, i::Int) = setindex!(a.A[i], x)
+for typ in Base.Threads.atomictypes
```

Comment:
This updates our `AtomicVector` type to be more efficient. Instead of storing a `Threads.Atomic{T}` for each element and having extra pointer indirections for each load/store, we generate the llvm intrinsic instructions ourselves for `getindex`/`setindex!` for each type.

Long term it'd be great to have a ThreadedDataStructures.jl package or something where this can live, but for now, this does what we want/need for multithreaded parsing, and it's pretty simple.

For reference, I basically copied the `llvmcall`s from Base and used [CudaNative.jl](https://github.com/JuliaGPU/CUDAnative.jl/blob/master/src/device/cuda/atomics.jl) and tips from @vchuravy/@maleadt.